File: data/graph-traversal/kruskals-algorithm.cpp
#include <vector>
#include <algorithm>

using namespace std;

class SolutionKruskal {
    struct Edge {
        int u, v, w;
        bool operator<(const Edge& other) const {
            return w < other.w;
        }
    };

    vector<int> parent;

    int find(int i) {
        if (parent[i] == i) return i;
        return parent[i] = find(parent[i]);
    }

    bool unionSets(int i, int j) {
        int rootI = find(i), rootJ = find(j);
        if (rootI != rootJ) {
            parent[rootI] = rootJ;
            return true;
        }
        return false;
    }

public:
    vector<vector<int>> kruskalsMST(int n, vector<vector<int>>& edgeList) {
        vector<Edge> edges;
        for (auto& e : edgeList) {
            edges.push_back({e[0], e[1], e[2]});
        }
        sort(edges.begin(), edges.end());

        parent.resize(n);
        for (int i = 0; i < n; ++i) parent[i] = i;

        vector<vector<int>> mst;
        for (auto& edge : edges) {
            if (unionSets(edge.u, edge.v)) {
                mst.push_back({edge.u, edge.v, edge.w});
            }
        }
        return mst;
    }
};

File: data/graph-traversal/kruskals-algorithm.py
def kruskalsMST(edges: list[list[int]], n: int) -> list[list[int]]:
    """
    Kruskal's Algorithm for Minimum Spanning Tree (MST)
    Time Complexity: O(E log E)
    Space Complexity: O(V)
    """
    edges.sort(key=lambda x: x[2])
    parent = list(range(n))

    def find(i):
        if parent[i] == i:
            return i
        parent[i] = find(parent[i])
        return parent[i]

    def union(i, j):
        root_i = find(i)
        root_j = find(j)
        if root_i != root_j:
            parent[root_i] = root_j
            return True
        return False

    mst = []
    for u, v, w in edges:
        if union(u, v):
            mst.append([u, v, w])
    return mst
